src/pybind/mgr/dashboard/services/auth.py

```diff
@@ -277,3 +280,92 @@ def _check_authorization(self, username):
         if not AuthManager.authorize(username, sec_scope, sec_perms):
             raise cherrypy.HTTPError(403, "You don't have permissions to "
                                           "access that resource")
+
+class Oauth2JWTManager:
+    _token = None
+    _token_payload = None
+    _user = None
+
+    @classmethod
+    def initialize_from_token(cls, token):
+        if not JwtManager.is_blocklisted(token):
+            cls._token = token
+            cls._token_payload = cls.get_token_payload()
+            cls._user = cls.get_user()
+
+    @classmethod
+    def reset(cls):
+        cls._token = None
+        cls._token_payload = None
+        cls._user = None
+
+    @classmethod
+    def get_token(cls):
+        return cls._token
+
+    @classmethod
+    def get_token_payload(cls):
+        if cls._token_payload:
+            return cls._token_payload
+        elif cls.get_token():
+            return decode_jwt_segment(cls._token.split(".")[1])
+        return {}
+
+    @classmethod
+    def get_token_from_header(cls):
+        try:
+            return cherrypy.request.cookie['token'].value
+        except KeyError:
+            return cherrypy.request.headers.get('X-Access-Token')
+
+    @classmethod
+    def get_user_roles(cls):
+        user_roles = []
+        if cls._token_payload:
+            # check for client roes
+            if 'resource_access' in cls._token_payload:
+                # Find the first value where the key is not 'account'
+                user_roles = Role.map_to_system_roles(next((value['roles'] for key, value in cls._token_payload['resource_access'].items() if key != "account"), user_roles))
+            # check for global roles
+            elif 'realm_access' in cls._token_payload:
+                user_roles = Role.map_to_system_roles(next((value['roles'] for _, value in cls._token_payload['realm_access'].items()), user_roles))
+            else:
+                raise DashboardException(f'Provided user roles, {user_roles} are not valid')
+        return user_roles
+
+    @classmethod
+    def get_user(cls):
+        if not cls._user:
+            cls._create_user()
+        return cls._user
+
+    @classmethod
+    def _create_user(cls):
+        if cls._token_payload:
+            try:
+                user = mgr.ACCESS_CTRL_DB.create_user(cls._token_payload['sub'], None, cls._token_payload['name'], cls._token_payload['email'])
+            except UserAlreadyExists:
+                user = mgr.ACCESS_CTRL_DB.get_user(cls._token_payload['sub'])
+            user.set_roles(cls.get_user_roles())
+            # set user last update to token time issued
+            user.last_update = cls._token_payload['iat']
+            cls._user = user
```

Comment:
I hate this, but I understand it's needed because the local JWT/Auth manager didn't set/read the roles from the JWT payload...